File: shared/src/main/java/de/dreier/mytargets/shared/targets/drawable/TargetDrawable.java
/*
 * Copyright (C) 2017 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.shared.targets.drawable;

import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Matrix;
import android.graphics.PixelFormat;
import android.graphics.PointF;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import de.dreier.mytargets.shared.models.Target;
import de.dreier.mytargets.shared.targets.models.TargetModelBase;
import de.dreier.mytargets.shared.targets.zone.ZoneBase;

public class TargetDrawable extends Drawable {

    public static final RectF SRC_RECT = new RectF(-1, -1, 1, 1);
    @NonNull
    protected final Target target;
    final TargetModelBase model;
    @NonNull
    private final List<ZoneBase> zonesToDraw;
    @NonNull
    private final ArrayList<Matrix> targetFaceMatrices;
    @NonNull
    private final ArrayList<Matrix> drawMatrices;
    private Matrix matrix = new Matrix();
    private float zoom = 1;
    private float px;
    private float py;
    private Matrix spotMatrix = new Matrix();
    private float xOffset;
    private float yOffset;
    @NonNull
    private CanvasWrapper canvasWrapper = new CanvasWrapper();

    public TargetDrawable(@NonNull Target target) {
        this.model = target.getModel();
        this.target = target;
        this.zonesToDraw = new ArrayList<>();
        for (int i = getZones() - 1; i >= 0; i--) {
            if (!model.shouldDrawZone(i, target.scoringStyle)) {
                continue;
            }
            zonesToDraw.add(model.getZone(i));
        }
        targetFaceMatrices = new ArrayList<>();
        drawMatrices = new ArrayList<>();
        for (int faceIndex = 0; faceIndex < model.getFaceCount(); faceIndex++) {
            targetFaceMatrices.add(calculateTargetFaceMatrix(faceIndex));
            drawMatrices.add(new Matrix());
        }
    }

    public void setMatrix(Matrix matrix) {
        this.matrix = matrix;
    }


    @Override
    protected void onBoundsChange(@NonNull Rect bounds) {
        RectF srcRectWithStroke = new RectF(SRC_RECT);
        final ZoneBase outerZone = model.getZone(model.getZoneCount() - 1);
        final float inset = -outerZone.getStrokeWidth() * 0.5f;
        srcRectWithStroke.inset(inset, inset);
        matrix.setRectToRect(srcRectWithStroke, new RectF(bounds), Matrix.ScaleToFit.CENTER);
    }

    @NonNull
    public Target getTarget() {
        return target;
    }

    private int getZones() {
        return model.getZoneCount();
    }

    @Override
    public void draw(@NonNull Canvas canvas) {
        canvasWrapper.setCanvas(canvas);
        for (int faceIndex = 0; faceIndex < model.facePositions.length; faceIndex++) {
            setMatrixForTargetFace(canvasWrapper, faceIndex);
            for (ZoneBase zone : zonesToDraw) {
                zone.drawFill(canvasWrapper);
            }
        }
        for (int faceIndex = 0; faceIndex < model.facePositions.length; faceIndex++) {
            setMatrixForTargetFace(canvasWrapper, faceIndex);
            for (ZoneBase zone : zonesToDraw) {
                zone.drawStroke(canvasWrapper);
            }
            onPostDraw(canvasWrapper, faceIndex);
        }
        canvasWrapper.releaseCanvas();
    }

    protected void setMatrixForTargetFace(@NonNull CanvasWrapper canvas, int faceIndex) {
        canvas.setMatrix(getTargetFaceMatrix(faceIndex));
    }

    private Matrix getTargetFaceMatrix(int faceIndex) {
        Matrix m = drawMatrices.get(faceIndex);
        m.set(getPreCalculatedFaceMatrix(faceIndex));
        m.postConcat(spotMatrix);
        m.postTranslate(-px - 1, -py - 1);
        m.postScale(zoom, zoom);
        m.postTranslate(zoom + px, zoom + py);
        m.postConcat(matrix);
        m.postTranslate(xOffset, yOffset);
        return m;
    }

    public Matrix getPreCalculatedFaceMatrix(int faceIndex) {
        return targetFaceMatrices.get(faceIndex);
    }

    @NonNull
    private Matrix calculateTargetFaceMatrix(int index) {
        PointF pos = model.facePositions[index % model.facePositions.length];
        Matrix matrix = new Matrix();
        matrix.setRectToRect(SRC_RECT,
                new RectF(pos.x - model.faceRadius, pos.y - model.faceRadius,
                        pos.x + model.faceRadius, pos.y + model.faceRadius),
                Matrix.ScaleToFit.FILL);
        return matrix;
    }

    protected void onPostDraw(CanvasWrapper canvas, int faceIndex) {
        if (model.getDecorator() != null) {
            model.getDecorator().drawDecoration(canvas);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (o instanceof TargetDrawable) {
            TargetDrawable t = (TargetDrawable) o;
            return t.target.id == target.id;
        }
        return false;
    }

    @Override
    public int getOpacity() {
        return PixelFormat.OPAQUE;
    }

    @Override
    public void setAlpha(int arg0) {
    }

    @Override
    public void setColorFilter(ColorFilter arg0) {
    }

    public void setZoom(float zoom) {
        this.zoom = zoom;
    }

    public void setMid(float px, float py) {
        this.px = px;
        this.py = py;
    }

    public void setOffset(float x, float y) {
        xOffset = x;
        yOffset = y;
    }

    public Matrix getSpotMatrix() {
        return spotMatrix;
    }

    public void setSpotMatrix(Matrix spotMatrix) {
        this.spotMatrix = spotMatrix;
    }
}

File: shared/src/main/java/de/dreier/mytargets/shared/targets/drawable/TargetDrawable.kt
/*
 * Copyright (C) 2017 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.shared.targets.drawable

import android.graphics.*
import android.graphics.drawable.Drawable
import de.dreier.mytargets.shared.models.Target
import de.dreier.mytargets.shared.targets.models.TargetModelBase
import de.dreier.mytargets.shared.targets.zone.ZoneBase
import java.util.*

open class TargetDrawable(val target: Target) : Drawable() {
    val model: TargetModelBase = target.model
    private val zonesToDraw: List<ZoneBase>
    private val targetFaceMatrices: ArrayList<Matrix>
    private val drawMatrices: ArrayList<Matrix>
    private var matrix = Matrix()
    private var zoom = 1f
    private var px: Float = 0f
    private var py: Float = 0f
    var spotMatrix = Matrix()
    private var xOffset: Float = 0f
    private var yOffset: Float = 0f
    private val canvasWrapper = CanvasWrapper()

    private val zones: Int
        get() = model.zoneCount

    init {
        this.zonesToDraw = (zones - 1 downTo 0)
                .filter { model.shouldDrawZone(it, target.scoringStyleIndex) }
                .map { model.getZone(it) }
        targetFaceMatrices = ArrayList()
        drawMatrices = ArrayList()
        for (faceIndex in 0 until model.faceCount) {
            targetFaceMatrices.add(calculateTargetFaceMatrix(faceIndex))
            drawMatrices.add(Matrix())
        }
    }

    fun setMatrix(matrix: Matrix) {
        this.matrix = matrix
    }

    override fun getIntrinsicHeight() = 2000

    override fun getIntrinsicWidth() = 2000

    override fun onBoundsChange(bounds: Rect) {
        val srcRectWithStroke = RectF(SRC_RECT)
        val outerZone = model.getZone(model.zoneCount - 1)
        val inset = -outerZone.strokeWidth * 0.5f
        srcRectWithStroke.inset(inset, inset)
        matrix.setRectToRect(srcRectWithStroke, RectF(bounds), Matrix.ScaleToFit.CENTER)
    }

    override fun draw(canvas: Canvas) {
        canvasWrapper.setCanvas(canvas)
        for (faceIndex in model.facePositions.indices) {
            setMatrixForTargetFace(canvasWrapper, faceIndex)
            for (zone in zonesToDraw) {
                zone.drawFill(canvasWrapper)
            }
        }
        for (faceIndex in model.facePositions.indices) {
            setMatrixForTargetFace(canvasWrapper, faceIndex)
            for (zone in zonesToDraw) {
                zone.drawStroke(canvasWrapper)
            }
            onPostDraw(canvasWrapper, faceIndex)
        }
        canvasWrapper.releaseCanvas()
    }

    private fun setMatrixForTargetFace(canvas: CanvasWrapper, faceIndex: Int) {
        canvas.setMatrix(getTargetFaceMatrix(faceIndex))
    }

    private fun getTargetFaceMatrix(faceIndex: Int): Matrix {
        val m = drawMatrices[faceIndex]
        m.set(getPreCalculatedFaceMatrix(faceIndex))
        m.postConcat(spotMatrix)
        m.postTranslate(-px - 1, -py - 1)
        m.postScale(zoom, zoom)
        m.postTranslate(zoom + px, zoom + py)
        m.postConcat(matrix)
        m.postTranslate(xOffset, yOffset)
        return m
    }

    fun getPreCalculatedFaceMatrix(faceIndex: Int): Matrix {
        return targetFaceMatrices[faceIndex]
    }

    private fun calculateTargetFaceMatrix(index: Int): Matrix {
        val pos = model.facePositions[index % model.facePositions.size]
        val matrix = Matrix()
        matrix.setRectToRect(SRC_RECT,
                RectF(pos.x - model.faceRadius, pos.y - model.faceRadius,
                        pos.x + model.faceRadius, pos.y + model.faceRadius),
                Matrix.ScaleToFit.FILL)
        return matrix
    }

    protected open fun onPostDraw(canvas: CanvasWrapper, faceIndex: Int) {
        model.decorator?.drawDecoration(canvas)
    }

    override fun getOpacity(): Int {
        return PixelFormat.OPAQUE
    }

    override fun setAlpha(arg0: Int) {}

    override fun setColorFilter(arg0: ColorFilter?) {}

    fun setZoom(zoom: Float) {
        this.zoom = zoom
    }

    fun setMid(px: Float, py: Float) {
        this.px = px
        this.py = py
    }

    fun setOffset(x: Float, y: Float) {
        xOffset = x
        yOffset = y
    }

    companion object {
        val SRC_RECT = RectF(-1f, -1f, 1f, 1f)
    }
}
